Migrate DQM to edm::Accumulator

schneiml opened the pull request:
This PR converts the entire DQM step1 to use `edm::one::EDProducer` modules with the new accumulate feature.

- The most interesting changes are in 53742cf.
- There is nothing actually produced in the new `EDProducer`s (no tokens etc.). The framework appears to be fine with that, thanks to the `edm::Accumulate` feature (thanks @wddgit !).
- Many more configurations had to be updated, see also #21871. More tests might uncover even more.
- Some useless overrides where removed to fix linking issues.
- So far, everything (even online DQM) seems to work. Therefore, there is no "opt out" for this conversion for now.
- The new `DQMGlobalEDAnalyzer`, however, is not affected, and might serve as an alternative where the `edm::one` approach does not fit (lots of work done in very few modules -> not enough tasks to keep all threads busy)
- If no opt-out is needed for anything, lots of code could be removed. This is not done yet.
- Update: removed some unneeded methods and fixed a few transitive includes. But there is still a large cleanup in `DQMStore` and `DQMEDAnalyzer` to be done, once we can get rid of streams.

If nobody discovers serious problems, this can be merged; otherwise, an opt-out or proper conversion of the respective workflows might be needed.

schneiml committed:
Remove thread_unsafe

schneiml committed:
Switch to edm::one::EDProducer, does basically compile

schneiml committed:
Name conflict with std::accumulate

schneiml committed:
Add more method dummies

schneiml committed:
Remove useless begin/endJob to fix linking problems

schneiml committed:
Migrate some forgotten configs

schneiml committed:
actually trigger the config switch

schneiml committed:
Manually migrate some more configs

schneiml committed:
Another round of automatic config changes...

... now for multi-line declarations in the config.

schneiml committed:
Convert some configs using templated plugin

schneiml committed:
More config fixes; standard sequence runs now

schneiml committed:
Switch back one config that was too much

schneiml committed:
Another config conversion

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22157/3294

cmsbuild commented:
A new Pull Request was created by  @schneiml (Marcel Schneider) for master.

It involves the following packages:

CalibTracker/SiStripChannelGain
DQM/EcalMonitorTasks
DQM/HcalTasks
DQM/L1TMonitor
DQM/Physics
DQM/SiStripMonitorHardware
DQM/SiStripMonitorPedestals
DQM/SiStripMonitorTrack
DQM/TrackerMonitorTrack
DQM/TrigXMonitor
DQMOffline/Alignment
DQMOffline/CalibCalo
DQMOffline/JetMET
DQMOffline/L1Trigger
DQMOffline/Trigger
DQMServices/Core
FastSimulation/MaterialEffects
HLTriggerOffline/Egamma
HLTriggerOffline/Exotica
Validation/RecoMET
Validation/RecoTau
Validation/RecoTrack


@ghellwig, @cmsbuild, @civanch, @vazzolini, @lveldere, @kmaeshima, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @nsmith-, @rekovic, @franzoni, @jfernan2, @cerminar, @thomreis, @vanbesien, @lpernie can you please review it and eventually sign? Thanks.
@ghellwig, @TaiSakuma, @felicepantaleo, @wmtford, @rappoccio, @argiro, @Martin-Grunewald, @fioriNTU, @thomreis, @ahinzmann, @threus, @mmusich, @seemasharmafnal, @mmarionncern, @kreczko, @hdelanno, @battibass, @makortel, @jhgoh, @jdolen, @jlagram, @HuguesBrun, @drkovalskyi, @OlivierBondu, @trocino, @rociovilar, @barvic, @DryRun, @GiacomoSguazzoni, @tocheng, @VinInn, @jdamgov, @nhanvtran, @gkasieczka, @rovere, @deguio, @schoef, @idebruyn, @ebrondol, @mtosi, @dgulhan, @gbenelli, @calderona, @matt-komm, @mariadalfonso, @folguera this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

schneiml commented:
@fwyzard and @Dr15Jones might be interested as well.

makortel commented on Validation/RecoTrack/python/MultiTrackValidator_cfi.py:
```diff
@@ -5,7 +5,8 @@
 
-multiTrackValidator = cms.EDAnalyzer(
+from DQMServices.Core.DQMEDAnalyzer import DQMEDAnalyzer
+multiTrackValidator = DQMEDAnalyzer(
```
Do I understand correctly that this PR does not change the behaviour of `DQMGlobalEDAnalyzer` in any way (i.e. those still need to be declared as `cms.EDAnalyzer`)? In that case, could you please leave MTV (also the "GenPs" variant above) out from this migration as #21794 already converts it to `DQMGlobalEDAnalyzer`?

schneiml replied:
Yes, `DQMGlobalEDAnalyzer` is not affected.
No, if this code does run in "normal" relval workflows, and is not converted yet, it would crash the tests.

So we have an unfortunate conflict here. #21871 was supposed to buffer that a bit, but as you see I missed some declarations there... In the worst case, I can factor out the config changes here and make another "noop" PR like #21871.

makortel replied:
Nah, let's treat this as a normal conflict then, i.e. whichever of the two gets merged last reacts.

fwyzard commented on CalibTracker/SiStripChannelGain/python/SiStripGainsPCLWorker_cfi.py:
```diff
@@ -1,7 +1,8 @@
-    "SiStripGainsPCLWorker",
+from DQMServices.Core.DQMEDAnalyzer import DQMEDAnalyzer
+SiStripGainsPCLWorker = DQMEDAnalyzer(
+    'SiStripGainsPCLWorker',
```
I would suggest not to change the `"` to `'`, to reduce the number of changes.

schneiml replied:
Artifact of the regexp used in my automated conversion. I can try to filter out these lines...

fwyzard commented on DQMServices/Core/interface/DQMEDAnalyzer.h:
```diff
@@ -18,63 +22,42 @@ namespace dqmDetails {struct NoCache {};}
+
+
+
+  /* (never called, only for stream-modules */ 
```
why do you keep methods that you say are "never called" ?

schneiml replied:
They should be removed, along with their implementation. However, I did not expect this migration to go as smoothly as it seems to be right now, and if we need a opt-out of some form, we still need them.

I'll either remove them later in this PR or in a follow-up.

schneiml committed:
Add correct override annotations

fwyzard commented:
> There is nothing actually produced in the new EDProducers (no tokens etc.). The framework appears to be fine with that, thanks to the edm::Accumulate feature.

As far as I know, the framework has no reason to complain if an `EDProducer` does not `produce` anything. However it will not be able to schedule the modules automatically, and they will need to be scheduled explicitly, just like the `EDAnalyzer`s.

So, I understand the reason for the migration from `stream` to `one`, but why the migration to `EDProducer` in the first place ?

fwyzard commented:
Regarding the migration from `stream` to `one` - what is the impact (if any) on the RECO+DQM step, running with 4 or 8 streams ?

schneiml commented:
@fwyzard  re the framework: Yes it is possible that it works right now because everything is still properly scheduled. However, my perception was that Producers _always_ get unscheduled and should not run at all when they are not needed, but this still works, since the accumulate feature keeps them in a separate collection. @Dr15Jones or @wddgit should probably answer this.

re timing (if that is what you are asking): it is faster when multithreaded, slightly slower in single threaded, in my simple tests. This roughly lines up with the exception: the unscheduled execution does cause some otherhead in any case, but with multiple threads, it is less than the cost of cloning/merging things.

schneiml committed:
Remove useless methods

fwyzard commented:
> @fwyzard  re the framework: Yes it is possible that it works right now because  everything is still properly scheduled. However, my perception was that  Producers always get unscheduled and should not run at all when  they are not needed, but this still works, since the accumulate feature  keeps them in a separate collection. @Dr15Jones or @wddgit should probably answer this.

If that is indeed the case, it should also work for analyzers.

> re timing (if that is what you are asking): it is faster when  multithreaded, slightly slower in single threaded, in my simple tests.

Mhm, both results are unexpected:
  - running single threaded, there should be no difference in the event loop
  - running multithreaded, there might be a small degradation because the framework cannot schedule the same `one` module at the same time

> This roughly lines up with the expection: the unscheduled execution does  cause some overhead in any case, 

No, it should not, because the scheduling is resolved before starting the event loop, based on the decalred dependencies.

> but with multiple threads, it is less  than the cost of cloning/merging things.

Yes, both single- and multithreaded there should be a performance gain in the endlumi/endrun transitions. Do you have some numbers ?

VinInn commented:
> 	• running multithreaded, there might be a small degradation because the framework cannot schedule the same one module at the same time
I think this is a major issue and we should really try to avoid such a sequentialization at module level.
There are use cases with few DQM modules and they will suffer.
We should really try to lock on single histo not module.

schneiml commented:
@VinInn there are few use cases with few modules, and these should be migrated to `DQMGlobalEDAnalyzer`. As long as it is only a few modules and they are not too legacy, I don't see a problem there.

Also, we are willing to accept some performance hit if we can get the memory savings instead. But it does not look like a performance hit at all for the common cases.

davidlange6 commented:
Some actual performance numbers would be of particular interest I think..

> On Feb 8, 2018, at 3:26 PM, Marcel Schneider <notifications@github.com> wrote:
> 
> @VinInn there are few use cases with few modules, and these should be migrated to DQMGlobalEDAnalyzer. As long as it is only a few modules and they are not too legacy, I don't see a problem there.
> 
> Also, we are willing to accept some performance hit if we can get the memory savings instead. But it does not look like a performance hit at all for the common cases.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Dr15Jones commented:
@fwyzard the acculator ability is only for EDProducers. This is because one can always put EDProducers on Tasks and get well defined behavior, even with accumulate. One can not put EDAnalyzers on Tasks. You can see the documentation at 

https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_Accumulator

Dr15Jones commented:
@fwyzard the migration to EDProducers is to allow the framework to see the true dependencies between modules. That way the OutputModule knows to run after all DQM modules are done and all the DQMEDHarversters know to run after the DQM modules that made the histograms which the harvesters are using.

fwyzard commented:
@Dr15Jones , well I disagree.

How can you expect "well defined behaviour" from a producer that does not `produce` anything ?
In fact, how is that different from an analyser ?

Dr15Jones commented:
@fwyzard the original agreement had them producing 'DQMTokens' at end Run and end Lumi.

schneiml commented:
@davidlange6 I have some numbers, but none that I consider reliable enough to circulate. I'll try to get better ones.

Also running the tests would be useful at some point (though I will have some cleanup commits in the next ~hour)

schneiml commented:
@fwyzard , @Dr15Jones yes, the original plan was (and is) to introduce the tokens, but since it seems to work without, this is a nice PR already.

That said, I don't really understed how it works atm, and I am only guessing (and inferring from the diff of #21920)

schneiml committed:
Fix modules relying on transitive includes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22157/3299

cmsbuild commented:
Pull request #22157 was updated. @ghellwig, @cmsbuild, @civanch, @vazzolini, @lveldere, @kmaeshima, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @nsmith-, @rekovic, @franzoni, @jfernan2, @cerminar, @thomreis, @vanbesien, @lpernie can you please check and sign again.

dmitrijus commented:
There are some 'old' benchmarks in https://twiki.cern.ch/twiki/bin/view/CMS/DQMFrameworkProjects#edm_one_Token_conversion_of_DQM

@schneiml will update the page after a fresh set of tests are done.

dmitrijus commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: a700068ef39acea17f78aa6cdb5f0712840b46f5

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22157/25979/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
7.3 step3
<pre>runTheMatrix-results/7.3_CosmicsSPLoose_UP17+CosmicsSPLoose_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17/step3_CosmicsSPLoose_UP17+CosmicsSPLoose_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17.log

</pre>136.731 step3
<pre>runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/step3_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2.log

</pre>136.788 step3
<pre>runTheMatrix-results/136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017/step3_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017.log

</pre>1325.7 step2
<pre>runTheMatrix-results/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017/step2_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017.log

</pre>1306.0 step3
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step3_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>1330.0 step3
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step3_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>10042.0 step3
<pre>runTheMatrix-results/10042.0_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step3_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>10024.0 step3
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>10224.0 step3
<pre>runTheMatrix-results/10224.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU.log

</pre>10824.0 step3
<pre>runTheMatrix-results/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018.log

</pre>11624.0 step3
<pre>runTheMatrix-results/11624.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2019_GenSimFull+DigiFull_2019+RecoFull_2019+ALCAFull_2019+HARVESTFull_2019/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2019_GenSimFull+DigiFull_2019+RecoFull_2019+ALCAFull_2019+HARVESTFull_2019.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

Dr15Jones commented:
The test failures appear to be because not all the python configuration files were changed to use `DQMEDAnalyzer` as the type. E.g. workflow 7.3 fails with
```
----- Begin Fatal Exception 08-Feb-2018 17:45:59 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=L1TMP7ZeroSupp label='l1tStage2BmtfZeroSupp'
Exception Message:
The base type in the python configuration is EDAnalyzer, but the base type
for the module's C++ class is EDProducer. Please fix the configuration.
It must use the same base type as the C++ class.
----- End Fatal Exception -------------------------------------------------
```

dmitrijus commented:
Updated performance: https://twiki.cern.ch/twiki/bin/view/CMS/DQMFrameworkProjects#edm_one_Token_conversion_of_DQM

This is very promising - cuts a large chunk of time wasted on histogram merging.

Note: this still 'merges' one local 'module' histogram into a 'global' one.
Eventually this will be improved (removed), but that requires a bit more work, after this pull request is merged.

schneiml commented:
@Dr15Jones yes, that is configs that I missed to convert. I'll update them.

schneiml commented:
@fwyzard I decided against using the Timing Service since I don't want to mess around with the real offline configs (which are huge). Getting sth. wrong there likely makes the results useless...

So I preferred a less useful, but actually correct measurement (or rather one that I trust).

Feel free to play with this PR if you have a test config for these things.

schneiml commented:
Re the output: By manual inspection it looks fine.

schneiml commented:
re the test failure: `l1tStage2BmtfZeroSupp` does not exist in the base IB. I have to rebase, it seems...

schneiml committed:
Fix a recently added config

cmsbuild commented:
The code-checks are being triggered in jenkins.

schneiml commented:
rebased, fixed the new config, and force-pushed. Good for another round of tests, I think.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22157/3304

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #22157 was updated. @ghellwig, @cmsbuild, @civanch, @vazzolini, @lveldere, @kmaeshima, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @nsmith-, @rekovic, @franzoni, @jfernan2, @cerminar, @thomreis, @vanbesien, @lpernie can you please check and sign again.

schneiml commented:
Ok, it appears DQM output *is* missing in some cases. Investigating...

fwyzard commented:
My worry is that the global end run transition seen by the one modules may happen after the DQMStore merges the per-instance plots into the global ones, so the DQM output module writes out empty plots.

Keep in mind though that I did not really investigate how the one module's transition line up, so it may be completely unrelated.

cmsbuild commented:
-1

Tested at: ab030dbe2b00363aebddc9d21e7f8bd10c58b45a

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22157/25988/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
1325.7 step2
<pre>runTheMatrix-results/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017/step2_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

schneiml committed:
Another config update

cmsbuild commented:
The code-checks are being triggered in jenkins.

schneiml commented:
Changed another declaration that made the tests fail. But it seems I can't run this wf locally, DAS does not return any results...

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22157/3311

cmsbuild commented:
Pull request #22157 was updated. @ghellwig, @cmsbuild, @civanch, @monttj, @vazzolini, @lveldere, @kmaeshima, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @nsmith-, @rekovic, @franzoni, @jfernan2, @cerminar, @thomreis, @vanbesien, @lpernie can you please check and sign again.

Dr15Jones commented:
So there was a question as to why this pull request is 'working'. The same question could be asked as to why the old code worked.

If a module (EDProducer, EDAnalyzer, etc) tells the framework it needs to see the Run and/or LuminosityBlock transitions then the framework will always call that module on that transitions, no matter what. The order that we call the modules is not guaranteed, unless one module puts data into the Run and/or LuminosityBlock and the other modules requests that data.

The Accumulator mechanism works similar, it tells the framework that the module absolutely needs to see ever Event in order to function. If the module is on a Path or EndPath then the user wanted the normal decision on calling the module to be made. If it is on a Task, the framework guarantees the module will be called for every Event. Note that if you use Accumulator then the EDProducer loses the 'produce' method for the Event, we should have enforced that the module produced something in either the Run or LuminosityBlock. Not enforcing that was an oversight.

So why did everything work in the past? The framework guarantees that all 'stream end' transitions must complete before any 'global end' transitions. Also, the OutputModules do not write a transition until all 'global end' calls are done. So in the past, the DQMEDAnalyzers were stream modules, the DQMEDHarvesters are one modules and so is MEtoEDMConverter. For this case, the framework made sure all DQMEDAnalyzers finish the Lumi or Run transition (including merging their histograms which occur on the stream transition) before DQMEDHarversters or MEtoEDMConverter are called (which are called on the global transition). Then the OutputModule is called after the global has been called. So everything worked EXCEPT MEtoEDMConverter did not work with DQMEDHarvester since they both were called on the global transition without the frameork knowing about any dependency between them.

So what is happening in this pull request? All DQMEDAnalyzers are being called on the global end transition. This means DQMHarvesters and DQMEDAnalyzers can not correctly be run in the same job. It also means MEtoEDMConverter does not work correctly since there is no guarantee the histograms have been finalized. The OutputModule is OK just because, for now, the framework guarantees the write happens after everything else. [I'd like to change that so we use proper data dependencies and so writes can run at the same time as unrelated activities on the end transition].

schneiml commented:
Thanks Chris,

explains a bit better what is going on there. What I don't understand at the moment is why this PR works fine single-threaded, but not multi-threaded. This might have sth. to do with whatever the DQMStore used to do internally (and still does, but should stop doing), but I don't see how it fits your explaination. Are we relying on "accidental ordering" of end run transitions there? Your remark on OutputModules seems to imply that this should work (even if not optimal).

I will have another deeper look at that next week, and we should have a chat on how to move forward there.

fwyzard commented:
> What I don't understand at the moment is why this PR works fine single-threaded, but not multi-threaded.

I think if you add the `Tracer` service to the test jobs, it should help you figure out the order in which the various signals are handled.

fwyzard commented:
By the way, if you ran `runTheMatrix.py -l 136.801`, the `DQMStore` is configured as
```python
process.DQMStore = cms.Service("DQMStore",
    LSbasedMode = cms.untracked.bool(False),
    collateHistograms = cms.untracked.bool(False),
    enableMultiThread = cms.untracked.bool(False),
    forceResetOnBeginLumi = cms.untracked.bool(False),
    referenceFileName = cms.untracked.string(''),
    verbose = cms.untracked.int32(0),
    verboseQT = cms.untracked.int32(0)
)
```

In particular **`enableMultiThread = cms.untracked.bool(False)`** .

However, running with multiple streams (e.g. by setting `process.options.numberOfStreams = cms.untracked.uint32(8)`) will automatically enable that flag; see https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/src/DQMStore.cc#L533_L537:
```c++
  ar.preallocateSignal_.connect([this](edm::service::SystemBounds const& iBounds) {
      if(iBounds.maxNumberOfStreams() > 1 ) {
	enableMultiThread_ = true;
      }
});
```

Setting `enableMultiThread_ = true` will alter the behaviour of the DQMStore and of the MonitorElements with respect to the end lumi and end run transitions.

schneiml committed:
Book all MEs as global

schneiml committed:
Call the endLumi transition for per lumi histograms.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22157/3358

cmsbuild commented:
Pull request #22157 was updated. @ghellwig, @cmsbuild, @civanch, @monttj, @vazzolini, @lveldere, @kmaeshima, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @nsmith-, @rekovic, @franzoni, @jfernan2, @cerminar, @thomreis, @vanbesien, @lpernie can you please check and sign again.

schneiml commented:
Rebased to pre1 and fixed two issues:
- Book all MEs as global. That was the reason why the output was not saved in MT, and @fwyzard 's hint regarding the `enableMultiThread` flag was on the right track.
- Call the end lumi processing in `DQMStore` to make sure the per-lumi histograms are produced. I am not sure if this is supposed to work for global MEs, but it seems to do the right thing for now.

In the long term:
- The `enableMultiThread` flag should disappear. There is no reason to have it if we only use `edm::global` and `edm::one` modules.
- The histogram merging code (that also saves the per-lumi histograms) needs to be redone, and mostly disappear.

I'd prefer to keep this PR minimal and only change what is needed to get the `edm::one` modules to work, and then have a follow-up to clean up things (which might involve some changes to the subsystem code, to get rid of deprecated functionality).

@fwyzard can you check (or tell me how to check) if/how this interacts with the `DQMGlobalEDAnalyzer`? I'd prefer it to not interact at all... But I heard there are issues with the per-lumi handling as well?

fwyzard commented:
@schneiml , from what I remember of the `DQMStore`, after the migration to `one` module the `DQMEDAnalyzer` should not need to call `mergeAndResetMEsLuminositySummaryCache(...)` and `mergeAndResetMEsRunSummaryCache(...)` any more (and you can simply get rid of the `endLuminosityBlock()` and `endRunSummary()` methods.

Those methods take care of merging the per-stream histograms into the global ones - but now all histograms are already global, so there is nothing to merge. In fact, reading the code for [`DQMStore::mergeAndResetMEsRunSummaryCache()`](https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/src/DQMStore.cc#L362_L435), it looks to me that you would end up duplicating the statistics of all histograms, because of [this call](https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/src/DQMStore.cc#L417) to
```
	me->getTH1()->Add(i->getTH1());
```

dmitrijus commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

fwyzard commented:
In fact, I would suggest to split the migration into multiple steps:
  - migrate from `stream` to `one` modules
  - migrate from `EDAnalyzer` to `EDProducer<Accumulator>`
  - introduce the per-lumi or per-run tokens

The first step should already find any problem in using global histograms and global transitions, and give you the memory saving.

The other steps take care of the migration from scheduled to unscheduled - but the two set of changes should be orthogonal.

schneiml commented:
@fwyzard yes, I am scared that something like that will happen. For now, calling `mergeAndResetMEsLuminositySummaryCache` with streamid 0 does produce the per-lumi copies (which is why I added it), but I did not follow thorugh the entire logic yet.

But after all, it *will* need some end lumi thing to make per-lumi histos work.

dmitrijus commented:
@fwyzard 

We have those global 'lumi' histograms, which 'artefacts' of merging and precisely same thing we have to delete at the post-end-lumi.

They are created by the merging - one copy for each lumi.

dmitrijus commented:
So for now, we keep creating this copies, which later can maybe be removed.

fwyzard commented:
> But after all, it will need some end lumi thing to make per-lumi histos work.

As long as you allow only one lumisection at a time, that should be already taken care of by the [`DQMStore::postGlobalBeginLumi()` method](https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/src/DQMStore.cc#L2154_L2180):
```c++
/** Called after all globalBeginLuminosityBlock.
 * Reset global per-lumi MEs (or all MEs if LSbasedMode) so that
 * they can be reused.
 */
void
DQMStore::postGlobalBeginLumi(const edm::GlobalContext &gc)
{
  ...
  while (i != e) {
    auto& me = const_cast<MonitorElement&>(*i++);
    // skip per-run MEs
    if (not LSbasedMode_ and not me.getLumiFlag())
      continue;
    me.Reset();
    me.resetUpdate();
  }
}
```

fwyzard commented:
> So for now, we keep creating this copies, which later can maybe be removed.

 no

schneiml commented:
@fwyzard oh, interesting. I did not notice `postGlobalBeginLumi` yet.

dmitrijus commented:
> If you read the code and the proposed modifications, you see that you are now calling mergeAndResetMEsLuminositySummaryCache() with stream id = 0 and module id = 0, so you are , merging the global histograms with themselves.

No.

The global histogram is seeded with lumisection number:
https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/src/DQMStore.cc#L470

So, technically, we are now merging histogram into their per-lumi 'snapshot' histogram.
Which is currently necessary, because we do need those 'per-lumi snapshots'.

I believe later these can be removed. However,  that disables parallel lumi processing, which is, as you said, already disabled by the fact those are one modules.

dmitrijus commented:
However, I agree that this needs to be removed/remorked.
Instead of full blown merging, we should simply inject those global per lumi snapshots. Which is only few lines of code.

schneiml commented:
@dmitrijus no, that is once again a different issue.

dmitrijus commented:
Ah, I see now.

fwyzard commented:
For a stepwise migration, what do you think of #22218 for the first step ?

Dr15Jones commented:
A step wise migration is not going to work if there are job where a DQMHarvester module is in the same configuration as the DQMEDAnalyzer to which it is harvesting. The reason is Paths/EndPaths are only used during Event transition. For Runs and Lumis all modules are run simultaneously with only data dependencies used to drive ordering of modules. The DQM just happens to work now since end stream transitions are always called for all modules before end global transitions.

fwyzard commented:
Hi Chris, I see.

However, a migration to `edm::one::Producer<edm::Accumulator>` without the introduction of run- or lumi-based tokens will have the same problem,  right ?

schneiml commented:
@Dr15Jones Afaik having a `DQMHarvester` in the same process as a `DQMEDAnalyzer` is not expected to do anything useful, except in online, which works differently anyways.

Dr15Jones commented:
> @Dr15Jones Afaik having a DQMHarvester in the same process as a DQMEDAnalyzer is not expected to do anything useful, except in online, which works differently anyways.

Yes. I believe I even said that somewhere in this pull request.

fwyzard commented:
> @fwyzard how is that different from switching to EDProducer<Accumulator>? Why would your PR work when adding the change to EDProducer<Accumulator> would fail?

As I wrote earlier, the two changes (from `stream` to `one` and from `EDAnalyzer` to `EDProducer<Accuulator>`) are almost orthogonal.

#22218 takes care of the first step.

The migration to `EDProducer` could be done on top of it.

schneiml commented:
@fwyzard I see, and I agree. Still, I don't really see not much of an advantage in separating out this, since that part actually works well so far. See my comment over there.

fwyzard commented:
indeed, feel free to take that commit and rebase the further migration on top of it.

fwyzard commented:
By the way, what is a Tier-0 like workflow with RECO and DQM to test these changes over 2017 data ?

schneiml commented:
This PR will be superseded by multiple others: #22243 for the config issues that were found here, and a variation of #22218 for the functional changes.
